cmd/util/ledger/migrations/multiple_contract_migration.go

```diff
@@ -0,0 +1,350 @@
+package migrations
+
+import (
+	"bytes"
+	"fmt"
+	"sort"
+	"strings"
+
+	"github.com/fxamacker/cbor/v2"
+	"github.com/onflow/cadence/runtime/ast"
+	"github.com/onflow/cadence/runtime/common"
+	"github.com/onflow/cadence/runtime/interpreter"
+	"github.com/onflow/cadence/runtime/parser2"
+	"github.com/rs/zerolog"
+	"github.com/rs/zerolog/log"
+
+	"github.com/onflow/flow-go/engine/execution/state"
+	"github.com/onflow/flow-go/ledger"
+	"github.com/onflow/flow-go/model/flow"
+)
+
+type MultipleContractMigrationError struct {
+	Errors []error
+}
+
+func (e *MultipleContractMigrationError) Error() string {
+	var b strings.Builder
+	b.WriteString("multiple contract migration errors:")
+	for i, err := range e.Errors {
+		_, err := fmt.Fprintf(&b, "\nerr %d: %s", i, err)
+		if err != nil {
+			panic(err)
+		}
+	}
+	return b.String()
+}
+
+var (
+	MultipleContractsSpecialMigrations = make(map[string]func(ledger.Payload) ([]ledger.Payload, error))
+)
+
+func MultipleContractMigration(payload []ledger.Payload) ([]ledger.Payload, error) {
+	migratedPayloads := make([]ledger.Payload, 0)
+	errors := make([]error, 0)
+
+	for _, p := range payload {
+		results, err := migrateRegister(p)
+		// dont fail fast... try to collect errors so multiple errors can be addressed at once
+		if err != nil {
+			errors = append(errors, err)
+			continue
+		}
+		migratedPayloads = append(migratedPayloads, results...)
+	}
+	if len(errors) != 0 {
+		return nil, &MultipleContractMigrationError{
+			Errors: errors,
+		}
+	}
+	return migratedPayloads, nil
+}
+
+func keyToRegisterId(key ledger.Key) (flow.RegisterID, error) {
+	if len(key.KeyParts) != 3 ||
+		key.KeyParts[0].Type != state.KeyPartOwner ||
+		key.KeyParts[1].Type != state.KeyPartController ||
+		key.KeyParts[2].Type != state.KeyPartKey {
+		return flow.RegisterID{}, fmt.Errorf("key not in expected format %s", key.String())
+	}
+
+	return flow.NewRegisterID(string(key.KeyParts[0].Value), string(key.KeyParts[1].Value), string(key.KeyParts[2].Value)), nil
+}
+
+func createContractNamesKey(originalKey ledger.Key) ledger.Key {
+	return ledger.Key{
+		KeyParts: []ledger.KeyPart{
+			originalKey.KeyParts[0],
+			originalKey.KeyParts[1],
+			{
+				Type:  state.KeyPartKey,
+				Value: []byte("contract_names"),
```

Comment:
Should maybe `state.keyContractNames` be used here?

Reply:
but thats private? should we export it for a single use migration?

Reply:
Ah, right! Yeah, not worth it to export for this one-off migration